### DISCUSSION TITLE
Does serenity.batch.count work for JUnit5?

### DISCUSSION THREAD

**Discussion by junjek** [serenity-bdd/serenity-core#3553] [Q&A]
I upgraded to 4.0 recently. In JUnit4, test executions can be split into batches in Jenkins. 
Example, 20 tests executed on 2 batches. The ideal expected result is batch1 executes 10tests and batch2 executes 10tests. The gist is tests are split and execute into batch. 

When I execute tests on JUnit5, I observed batch 1 run all 20 tests and batch 2 run all 20 tests with similar set up.

Therefore, would like to confirm if batch is working for JUnit5 or is there additional set up I need to perform?

**Comment by wakaleo** [comment]
I don't believe batching is implemented in the JUnit 5 runners. @nbarrett might know  more.

**Reply by junjek** [reply]
Thanks for the quick reply. 
Hello @nbarrett , could you kindly confirm that batching is not implemented in JUnit5 runner? If not, will there be future plan for this? It will help me to plan ahead. Thanks!

**Reply by nbarrett** [reply]
Hello, On my client project we did implement a batching mechanism for Junit 5 that uses SERENITY_BATCH_COUNT, SERENITY_BATCH_NUMBER etc but it was not contributed back to the project. I'll see whether I can create a standalone PR for this and reply back soon 👍

**Reply by junjek** [reply]
Thank you! Look forward to your update. :)

**Reply by junjek** [reply]
Hello @nbarrett , how are you. Would like to check if the batching mechanism has been implemented? Thanks.

**Reply by nbarrett** [reply]
Hello - sorry for delay - I have a changeset on my machine and will try to push it in the next day or so 🤞

**Reply by junjek** [reply]
Hi @nbarrett , would like to check if the batching feature has been pushed? Thank you.

**Reply by nbarrett** [reply]
Hello, sorry for late response. I plan to do this tomorrow 🤞

**Reply by nbarrett** [reply]
Update @junjek  💡 - I've just opened https://github.com/serenity-bdd/serenity-core/pull/3568 - take a look and see what you think 🤞

**Reply by junjek** [reply]
Hello @nbarrett @wakaleo ,
Will there be any plan to implement batching to Junit5 since Junit4 also have this feature? 
I tested v4.2.8. It seems batching doesn't seem to work on Junit5.

I also tried on cucumber, but runner couldn't execute test with "cucumber-batch" engine. Below is the log I received. It seems "cucumber-batch" is not a registered TestEngine.

> Internal Error occurred.
> org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-platform-suite' failed to discover tests
> at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:160)
> at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:132)
> at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:107)
> at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:78)
> at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:99)
> at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:85)
> at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
> at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:63)
> at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
> at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
> at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
> at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
> at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
> at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
> Caused by: org.junit.platform.commons.JUnitException: No TestEngine ID matched the following include EngineFilters: [cucumber-batch].
> Please fix/remove the filter or add the engine.
> Registered TestEngines:
> 
> junit-jupiter (group ID: org.junit.jupiter, artifact ID: junit-jupiter-engine, version: 5.11.1, location: jar:file:/C:/Repository/org/junit/jupiter/junit-jupiter-engine/5.11.1/junit-jupiter-engine-5.11.1.jar!/org/junit/jupiter/engine/JupiterTestEngine.class)
> junit-platform-suite (group ID: org.junit.platform, artifact ID: junit-platform-suite-engine, version: 1.11.1, location: jar:file:/C:/Repository/org/junit/platform/junit-platform-suite-engine/1.11.1/junit-platform-suite-engine-1.11.1.jar!/org/junit/platform/suite/engine/SuiteTestEngine.class)
> cucumber (version: DEVELOPMENT, location: jar:file:/C:/Repository/io/cucumber/cucumber-junit-platform-engine/7.20.1/cucumber-junit-platform-engine-7.20.1.jar!/io/cucumber/junit/platform/engine/CucumberTestEngine.class)
> junit-vintage (group ID: org.junit.vintage, artifact ID: junit-vintage-engine, version: 5.11.1, location: jar:file:/C:/Repository/org/junit/vintage/junit-vintage-engine/5.11.1/junit-vintage-engine-5.11.1.jar!/org/junit/vintage/engine/VintageTestEngine.class)
> Registered EngineFilters:
> EngineFilter that includes engines with IDs [cucumber-batch]
> at org.junit.platform.launcher.core.EngineFilterer.checkNoUnmatchedIncludeFilter(EngineFilterer.java:67)
> at org.junit.platform.launcher.core.EngineFilterer.performSanityChecks(EngineFilterer.java:51)
> at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:136)
> at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:107)
> at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:97)
> at org.junit.platform.suite.engine.SuiteLauncher.discover(SuiteLauncher.java:57)
> at org.junit.platform.suite.engine.SuiteTestDescriptor.discover(SuiteTestDescriptor.java:105)
> at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
> at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
> at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
> at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
> at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
> at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
> at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
> at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
> at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
> at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
> at org.junit.platform.suite.engine.DiscoverySelectorResolver.discoverSuites(DiscoverySelectorResolver.java:36)
> at org.junit.platform.suite.engine.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:42)
> at org.junit.platform.suite.engine.SuiteTestEngine.discover(SuiteTestEngine.java:58)
> at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
> ... 13 more
> Process finished with exit code -2

**Comment by timothyjmtan** [comment]
Good morning @nbarrett, @wakaleo, would you be able to take a look at this request again? Seems like Nick did a PR in November but it's been tested that batching is still not working (v4.2.8).
